feature: geoip stage in promtail (#3493)
grafana/loki: docs/sources/clients/promtail/stages/geoip.md

---
title: geoip
description: geoip stage
---
# geoip

> The `geoip` stage is a parsing stage that reads a ip address and 

Reviewer: ```suggestion
The `geoip` stage is a parsing stage that reads an IP address and 
```

populates the labelset with geoip fields. Maxmind's GeoIP2 databse is used for the lookup.

> Populated fields for City db:

Reviewer: Question, in line 8 it says this is using "GeoIP2 database", and in line 10 it says "City database".    Which is correct?  Could we clarify or standardize the terminology used for the database?

Author: The general term used is `GeoIP2 database` but is has many types of DB's including `City`, `ASN` and `Country`.  Lines `10` and `23` indicates the same with corresponding documentation.

- geoip_city_name
- geoip_country_name
- geoip_continet_name
- geoip_continent_code
- geoip_location_latitude
- geoip_location_longitude
- geoip_postal_code
- geoip_timezone
- geoip_subdivision_name
- geoip_subdivision_code

> Populated fields for ASN db:

Reviewer: And a third database mention.  Also, please spell out acronyms (Acronym Finder has 60+ entries for ASN).

Author: ASN is `Autonomous System Number`. Since, we are speaking of IPs here, it's more obvious. So, I feel it's okay not to expand it.

Reviewer: As a rule, always spell out acronyms.  Don't assume that just because you know what the acronym means that others will also know. Spelling out acronyms is helpful for non-native English speakers, and absolutely essential for translation/localization purposes.

- geoip_autonomous_system_number
- geoip_autonomous_system_organization

## Schema

```yaml
geoip:
  # Path to the Maxmind DB file
  [db: <string>]

  # IP from extracted data to parse.
  [source: <string>]
  
  # Maxmind DB type. Allowed values are "city", "asn"
  [db_type: <string>]
```

> ## GeoIP with City database

Reviewer: ```suggestion
## GeoIP with City database example
```

For the given pipeline

```yaml
- regex:
    expression: "^(?P<ip>\S+) .*"
- geoip:
    db: "/path/to/GeoIP2-City.mmdb"
    source: "ip"
    db_type: "city"
```

And the log line:

```
"34.120.177.193 - "POST /loki/api/push/ HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"
```

> The `regex` stage parses the log line and `ip` is extracted. Then the extracted `ip` value is given as `source` to `geoip` stage. `geoip` stage performs a lookup on the `ip` and populates below labels

Reviewer: ```suggestion
The `regex` stage parses the log line and `ip` is extracted. Then the extracted `ip` value is given as `source` to `geoip` stage.  The`geoip` stage performs a lookup on the `ip` and populates the following labels:
```

- `geoip_city_name`: `Kansas City`
- `geoip_country_name`: `United States`
- `geoip_continet_name`: `North America`
- `geoip_continent_code`: `NA`
- `geoip_location_latitude`: `"39.1027`
- `geoip_location_longitude`: `-94.5778`
- `geoip_postal_code`: `64184`
- `geoip_timezone`: `America/Chicago`
- `geoip_subdivision_name`: `Missouri`
- `geoip_subdivision_code`: `MO`

> If only a subset of above labels are required. We can chain the above pipeline with `labeldrop` or `labelallow` stage

Reviewer: ```suggestion
If only a subset of these labels are required, you can chain the above pipeline with the `labeldrop` or `labelallow` stages.
```

> ### labelallow

Reviewer: ```suggestion
### labelallow example
```

```yaml
- regex:
    expression: "^(?P<ip>\S+) .*"
- geoip:
    db: "/path/to/GeoCity.mmdb"
    source: "ip"
    db_type: "city"
- labelallow:
  - geoip_city_name
  - geoip_country_name
  - geoip_location_latitude
  - geoip_location_longitude
```

Only the labels listed under `labelallow` will be sent to Loki.

> ### labeldrop

Reviewer: ```suggestion
### labeldrop example
```

```yaml
- regex:
    expression: "^(?P<ip>\S+) .*"
- geoip:
    db: "/path/to/GeoCity.mmdb"
    source: "ip"
    db_type: "city"
- labeldrop:
  - geoip_postal_code
  - geoip_subdivision_code
```

All the labels except the ones listed under `labeldrop` will be sent to Loki.

> ## GeoIP with ASN database

Reviewer: ```suggestion
## GeoIP with ASN database example
```

```yaml
- regex:
    expression: "^(?P<ip>\S+) .*"
- geoip:
    db: "/path/to/GeoIP2-ASN.mmdb"
    source: "ip"
    db_type: "asn"
```

And the log line:

```
"34.120.177.193 - "POST /loki/api/push/ HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"
```

> The `regex` stage parses the log line and `ip` is extracted. Then the extracted `ip` value is given as `source` to `geoip` stage. `geoip` stage performs a lookup on the `ip` and populates below labels

Reviewer: ```suggestion
The `regex` stage parses the log line and `ip` is extracted. Then the extracted `ip` value is given as `source` to the `geoip` stage.  The `geoip` stage performs a lookup on the `ip` and populates the following labels:
```
